packages/dds/tree/docs/wip/Container and Application Schema.md

```diff
@@ -0,0 +1,194 @@
+
+# Expected usage patterns
+
+I suspect different users will mostly fall into 2 groups:
+
+-   pure schema-on-read: no customization of container schema.
+-   almost pure schema-on-write: only using schema-on-read to assist with schema migration so their application schema doesn't have to deal with old formats (which unavoidably pile up in the container schema)
+    They may additionally use schema-on-read to enforce/fix some invariants the container schema system isn't expressive enough to declare (referential integrity for graph like references, numeric ranges, and any other appellation level invariants that are desired)
+
+# Example/Proposed Design
+
+This design shows an easy to implement MVP, and is not intended to have document or API compatibility with what we do longer term.
+It minimally implements the requirements from "What options to support" while being extensible in the future.
+It can be subseted for use in earlier milestones, and could be extended for later ones, see "Schedule" below.
+
+This design is intentionally minimal while still being useful.
+Features not needed for minimal strongly typed application use or the JSON and XML domain schema are not included.
+
+## Schema Representation
+
+Include a simple non-contextual schema system, usable as a MVP for both `container schema` and `application schema`.
+
+We can start with:
+
+```typescript
+enum Multiplicity {
+    Value,
+    Optional,
+    Sequence,
+}
+
+enum Value {
+    Nothing,
+    Number,
+    String,
+    Boolean,
+    Serializable,
+}
+
+interface FieldContent {
+    multiplicity: Multiplicity;
+    types?: Type[];
+}
+
+interface Field {
+    content: FieldContent;
+    name: string;
+}
+
+interface Type {
+    name: string;
+    fields: Field[];
+    extraFields: FieldContent;
+    value: Value;
+}
+```
+
+These `Types` can be added as `container schema` as part of an edit op, which is considered conflicted if it tries to add a type that has a conflicting `name` and is not equal to the existing one.
+
+This intentionally does not support any concept of versioning or name-spacing: users of it can include versions and namespaces in the `name` if they want, or features for them could be added later.
+
+While a schema must be added to the document as a container schema to use the type
+(otherwise adding a new type could break existing data which might not even be downloaded on the current client),
+it's possible to add a schema that is compatible with all possible data.
+Applications which wish to rely entirely on schema-on-read for some or all of their data can use this pattern for all `container schema` and only use their actual developer authored schema as `application schema`:
+
+```typescript
+const anyField: FieldContent = { multiplicity: Multiplicity.sequence };
+function makeAnyType(name: string): Type {
+    return { name, extraFields: anyField };
+}
+```
```

Comment:
FieldContent.types is an array, and thus a union.

Note that this function creates a type that allows any children. If you instead want to declare that a field permits any type, you specify its `types` as `undefined`.

If you want a field that can contain any Json value, then you use the array of all Json types. I have added Json as an example domain to help show this.

Reply:
Did you consider making union a type constructor, rather than fixing it to the structure of node field types? I could see people wanting to say that a field satisfies multiple types (also via a list). Is primarily to unblock the initial json case?

Reply:
makeAnyType was intended as an example to show how flexible types could be expressed.
It is no longer present in the updated code since names have been factored apart from the `Type` (which has been renamed to TreeSchema`. This allows `anyTree` to exist as a constant (which is in the code now), which replaces makeAnyType.

@jack-williams I'm a bit confused by your question. All trees currently have a type explicitly on them, and our schema system only refers to those types. There is no way (currently in this example) to apply structural constraints to children, you can only constraint what types of children are allowed. All types on fields are done via a tagged union type constructor. If you wanted to be explicit you could do:
```typescript
export type TaggedUnion = ReadonlySet<TreeSchemaIdentifier>;
export function union(
    ...types: TreeSchemaIdentifier[]
): TaggedUnion {
    return new Set(types);
}
```
I'm not sure what you mean by "a field satisfies multiple types".

Anyway, the example code has changed significantly since your comment, and there are many more comments in code now: let me know if there is still something that is unclear, or if you still have questions about this.

Reply:
@jack-williams - Just FYI - Craig answered my question, but I'll leave this unresolved in case you wanted to respond regarding unions.